### PR TITLE
Support "group:artifact" notation in Version Catalogs

### PR DESCRIPTION
**Pull request by martinbonnin**
In version catalogs, it is currently not possible to omit the version from "String" notations:

```
# Not working 
compose-icons = "androidx.compose.material:material-icons-extended"
# Working
compose-icons = { module = "androidx.compose.material:material-icons-extended" }
```

This PR fixes this and also for the `settings.gradle` DSL:

```kotlin
  versionCatalogs {
    create("libs") {
      // working now
      library("compose.icons", "androidx.compose.material:material-icons-extended")
    }
  }
```

Fixes https://github.com/gradle/gradle/issues/24169

### Context
Slack discussion: [linen](https://linen.dev/s/gradle-community/t/9687207/minor-annoyance-with-version-catalogs-coordinates-without-a-#a89712a3-98ff-4480-9e45-e5cdd8675e55)

### Contributor Checklist
- [x] [Review Contribution Guidelines](https://github.com/gradle/gradle/blob/master/CONTRIBUTING.md)
- [x] Make sure that all commits are [signed off](https://git-scm.com/docs/git-commit#Documentation/git-commit.txt---signoff) to indicate that you agree to the terms of [Developer Certificate of Origin](https://developercertificate.org/).
- [x] Make sure all contributed code can be distributed under the terms of the [Apache License 2.0](https://github.com/gradle/gradle/blob/master/LICENSE), e.g. the code was written by yourself or the original code is licensed under [a license compatible to Apache License 2.0](https://apache.org/legal/resolved.html).
- [x] Check ["Allow edit from maintainers" option](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/) in pull request so that additional changes can be pushed by Gradle team
- [x] Provide integration tests (under `<subproject>/src/integTest`) to verify changes from a user perspective
- [x] Provide unit tests (under `<subproject>/src/test`) to verify logic
- [ ] Update User Guide, DSL Reference, and Javadoc for public-facing changes
- [x] Ensure that tests pass sanity check: `./gradlew sanityCheck`
- [x] Ensure that tests pass locally: `./gradlew <changed-subproject>:quickTest`

### Gradle Core Team Checklist
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation

### CODE REVIEW

File: subprojects/core/src/main/java/org/gradle/api/internal/catalog/parser/TomlCatalogFileParser.java [hunk 1]
```diff
@@ -254,12 +254,21 @@ private static void parseLibrary(String alias, TomlTable librariesTable, Version
         Object gav = librariesTable.get(alias);
         if (gav instanceof String) {
             List<String> splitted = SPLITTER.splitToList((String) gav);
-            if (splitted.size() == 3) {
+            int splittedSize = splitted.size();
+            if (splittedSize >= 2 && splittedSize <= 3) {
```
Comment (oldergod) [inline subprojects/core/src/main/java/org/gradle/api/internal/catalog/parser/TomlCatalogFileParser.java:258]:
```suggestion
            if (splittedSize == 2 && splittedSize == 3) {
```

File: subprojects/dependency-management/src/main/java/org/gradle/api/internal/catalog/DefaultVersionCatalogBuilder.java [hunk 2]
```diff
@@ -324,8 +324,13 @@ public void library(String alias, String groupArtifactVersion) {
         String normalizedAlias = normalizeAndValidateAlias(AliasType.LIBRARY, alias);
 
         String[] coordinates = groupArtifactVersion.split(":");
-        if (coordinates.length == 3) {
-            objects.newInstance(DefaultLibraryAliasBuilder.class, DefaultVersionCatalogBuilder.this, normalizedAlias, coordinates[0], coordinates[1]).version(coordinates[2]);
+        if (coordinates.length >= 2 && coordinates.length <= 3) {
```
Comment (oldergod) [inline subprojects/dependency-management/src/main/java/org/gradle/api/internal/catalog/DefaultVersionCatalogBuilder.java:327]:
```suggestion
        if (coordinates.length == 2 && coordinates.length == 3) {
```

**Review by oldergod — APPROVED** [review]
Let's go

File: subprojects/core/src/main/java/org/gradle/api/internal/catalog/parser/TomlCatalogFileParser.java (hunk 1 above)
Comment (martinbonnin) [reply subprojects/core/src/main/java/org/gradle/api/internal/catalog/parser/TomlCatalogFileParser.java:258]: You mean that with a `||`, right? 😄

Comment (oldergod) [reply subprojects/core/src/main/java/org/gradle/api/internal/catalog/parser/TomlCatalogFileParser.java]: yes ! Thank you
Way simpler to read I thought

Comment (martinbonnin) [reply subprojects/core/src/main/java/org/gradle/api/internal/catalog/parser/TomlCatalogFileParser.java]: 👍 change is pushed

File: subprojects/dependency-management/src/test/groovy/org/gradle/api/internal/catalog/DefaultVersionCatalogBuilderTest.groovy [hunk 3]
```diff
@@ -51,7 +51,7 @@ class DefaultVersionCatalogBuilderTest extends Specification implements VersionC
             alias('foo')
         })
         where:
-        notation << ["", "a", "a:", "a:b", ":b", "a:b:", ":::", "a:b:c:d"]
+        notation << ["", "a", "a:", ":b", ":::", "a:b:c:d"]
```
Comment (martinbonnin) [inline subprojects/dependency-management/src/test/groovy/org/gradle/api/internal/catalog/DefaultVersionCatalogBuilderTest.groovy:54]: Should `"a:b:"` be an error?

Comment (oldergod) [reply subprojects/dependency-management/src/test/groovy/org/gradle/api/internal/catalog/DefaultVersionCatalogBuilderTest.groovy:54]: yes

Comment (martinbonnin) [reply subprojects/dependency-management/src/test/groovy/org/gradle/api/internal/catalog/DefaultVersionCatalogBuilderTest.groovy:54]: I don't know, this works perfectly:

```
dependencies {
  implementation("com.google.firebase:firebase-perf-ktx:")
}
```

**Comment by martinbonnin** [comment]
Any chances this could land in `8.1`?

**Comment by eskatos** [comment]
@martinbonnin  8.1 ship has sailed as RC1 is already out

**Comment by martinbonnin** [comment]
Damn! Alright, thanks a lot for the update! Looking forward to 8.2 :)

**Comment by martinbonnin** [comment]
Hi 👋 anything holding this PR?

**Comment by big-guy** [comment]
We don't want to add another notation, but we can document provide better hints at what the correct notation is in the error.

**Comment by martinbonnin** [comment]
The notation is there in [DependencyHandler.create](https://docs.gradle.org/current/javadoc/org/gradle/api/artifacts/dsl/DependencyHandler.html#create-java.lang.Object-) already. 

I find it really annoying that I can't copy/paste between `build.gradle` and version catalogs. Is there a reason this is not allowed?

**Comment by jjohannes** [comment]
I understand the general argument of "we do not want (too many) alternative notations". But I don't agree that it applies in this case because:
- There is already the alternative notation for full GAV: `group:artifact:version` works
- `group:artifact` is a natural variation of that
- Furthermore, I would argue `group:artifact` is even THE best notation in a consistent resolution setup where you want to separate coordinates – addressing the "component" - from the version - defining "versions constraints". (In my opinion, this should be the primary advertised notation for dependency declaration.)

Furthermore, implementing this is really no hassle as the notation is already supported by all underlying systems, as @martinbonnin shows in this PR

**Comment by LouisCAD** [comment]
Can this be reopened, then?

**Pull request closed without merging**